src/pymoca/backends/casadi/model.py

```diff
@@ -568,21 +580,61 @@ def extract_assignment(eq):
             variables = []
             values = []
 
+            def get_derivative(expr):
+                if expr.is_constant():
+                    return 0.0
+                elif expr.is_symbolic():
+                    if expr.name() in states:
+                        return der_states[list(states.keys()).index(expr.name())].symbol
+                    elif expr.name() in alg_states:
+                        # This algebraic state must now become a differentiated state.
+                        states[expr.name()] = alg_states.pop(expr.name())
+                        der_sym = ca.MX.sym('der({})'.format(expr.name()))
+                        der_states.append(Variable(der_sym, float))
+                        return der_sym
+                    else:
+                        return 0.0
+                else:
+                    # Differentiate using CasADi and chain rule
+                    orig_deps = ca.symvar(expr)
+                    deps = ca.vertcat(*orig_deps)
+                    J = ca.Function('J', [deps], [ca.jacobian(expr, deps)])
+                    J_sparsity = J.sparsity_out(0)
+                    der_deps = [get_derivative(dep) if J_sparsity.has_nz(0, j) else ca.DM.zeros(dep.size()) for j, dep in enumerate(orig_deps)]
+                    return ca.mtimes(J(deps), ca.vertcat(*der_deps))
+
             reduced_equations = []
             for eq in self.equations:
                 variable, value = extract_assignment(eq)
                 if variable is not None:
+                    if variable.name() in states:
+                        # Mark derivative state for replacement too.
+                        derivative = get_derivative(value)
+
+                        state_index = list(states.keys()).index(variable.name())
```

Comment:
Using `.index` could be dangerously slow, and will likely come back to bite us. It's probably better to make a dict `MX state -> MX derivative` or `str state (name) -> MX derivative`. Note that all other lookups like `eq.dep(0).name() in states` are already O(1).